Feature/ceiroa/tdtm advancement

weizsforg opened the pull request:
The purpose of this PR is to implement the dependency injection mechnism between NPSP and Gift Entry. The idea is that the NPSP will get the TDTM Trigger dynamically based on if Gift Entry is installed or not.

Implemented the dependency injection mechnism for Gift Entry. So, Gift Entry can return a list of TiggerHanders for NPSP TDTM framework only when Gift Entry is installed.  Also, disabled some unit test cases when Gift Entry is installed. 

*** Note to how to test the PR. There shouldn't be any logic changes in NPSP existing functionality. QA only needs to do the regression test for this PR and make sure the existing test cases pass.

# Critical Changes

# Changes
- Implemented internal TDTM enhancements to support future development.

# Issues Closed
Fixes #2995
# New Metadata
Apex class: 
ADV_Adapter
ADV_Info
ADV_Adapter_TEST

Custom Label: 
adapterException
advConfigException
# Deleted Metadata

davidhabib commented:
@weizsforg @ceiroa When opening a pull request in NPSP, it would be good to provide more details on what this is for and whether it is ready above the header comments (so it won't be included in release notes), and assign it to me (I may delegate later).  Also, we then need the appropriate public comments added in the release notes section.

For this pull request specifically, is this final and truly ready?  I'm not a big fan of calling the new global class TDTM_Adapter if the intention is for it to encapsulate additional communication between NPSP and Adv that has nothing to do with TDTM.  Not sure what your current plans are for this class, so I'd like to hear more.  Once its global, we have to live with it forever! thanks.

weizsforg commented:
@davidhabib Thanks for the comments. I can add some description for this PR. Currently, we are still working on it. So, it is not ready for review yet. After we finalize it, we will assign it to you for review. 

We are going to change TDTM_Adapter to a appropriate name later since I saw your comments in HEDA branch. And we can talk about it more when Carlos back next week. 

Please add comments on this PR so we can address it when we update the TDTM_Adapter class name.

weizsforg commented on src/classes/TDTM_ObjectDataGateway.cls:
```diff
@@ -91,9 +91,36 @@ public without sharing class TDTM_ObjectDataGateway implements TDTM_iTableDataGa
+                // HEDA's TDTM config.
+            }
+
             if (listTH == null) {
```
Here we say if the Trigger Handler is empty, then we try to use Adv version of Trigger Handler. But I think we also need to take care of when listTH is not empty. We still need Adv to overwrite the list of Trigger Handlers.

ceiroa commented on src/classes/Advancement_Adapter.cls:
```diff
@@ -0,0 +1,59 @@
+     *
+     * @return none
+     */
+    global static void rollupAccounts(list<Id> listAccountId) {
```
@force2b @njjc is asking to let you know Advancement needs a global method for this, and if you'd be interested in creating it.

ceiroa replied:
Class would be called RLLP_Rollup_API or something similar.

ceiroa replied:
@force2b Hey Mike, just following up on this. Do you guys see this coming out any time soon?

davidhabib replied:
I'll add an AA item for this and put in our upcoming sprint so it gets discussed tomorrow.

weizsforg replied:
Hi @davidhabib , thanks. 

Other than the rollup functionality, Advancement also need other logic from NPSP. The logic is about the moveOppsToContactAccount method in ACCT_IndividualAccounts_TDTM class.  Can we also have an API for this part of the logic?

njjc replied:
We've decided to move forward with including this rollup method and the other methods you need in this class, instead of trying to define good externally consumable global APIs in time for this project. Let's just include the logic to make these methods throw an exception if the ADV package isn't installed.

ceiroa commented on src/classes/Advancement_Info.cls:
```diff
@@ -0,0 +1,77 @@
+    private static void setup() {
+        //Check if Advancement is installed and enabled.
+        try {
+            List<SObject> advSettings = Database.query('select Use_Advancement__c, API_NPSP_Class__c from Advancement_Setting__mdt');
```
We need to add the Advancement prefix of the final packaging org.

ceiroa replied:
We should first check that Advancement is actually installed through:

    System.isCurrentUserLicensed('mynamespace');

njjc replied:
https://developer.salesforce.com/forums/?id=906F000000092XIIAY

ceiroa replied:
Temporary Advancement prefix is adv1.

weizsforg commented on src/classes/Advancement_Adapter.cls:
```diff
@@ -0,0 +1,59 @@
+* @description Provides an adapter for Advancement package to be able to inject customizations into NPSP and provide API to inject to other project.
+**/
+
+global virtual with sharing class Advancement_Adapter {
```
Throw exception in all of the methods if Adv is not installed

weizsforg commented:
Thanks @davidhabib @njjc @ceiroa @pedrodalcol for your time! And thanks @davidhabib @njjc for the great feedback about this PR! I will clean up this PR and submit for review soon.

weizsforg commented:
@lianeyee For the Advancement project, we need an error message for one of the exception. 
The use case is that we have some global api for Advancement in NPSP. And we do not want other customer to call it by accident. 
I am thinking to throw an error like 'This function is for Advancement project, please do not call this function!' Is this message ok?
Please note that the message is for Apex Exception, it will not be in the UI.

lianeyee commented:
@weizsforg Could we say this? "Invalid call. This function is reserved for Salesforce use."

weizsforg commented:
@lianeyee Looks great! Thanks

lianeyee commented:
@juliesnow Testing whether you can see this?

juliesnow commented:
Yep...I see Review required, All checks have passed, and Merging is blocked. Is that what you expect me to be able to see?

juliesnow commented:
And, of course, the whole thread between you and Wei

lianeyee commented:
Great! Not sure why when I mention your username @juliesnow , I don't get the prompt that I usually get for other users. But glad that you're officially in the loop!

weizsforg commented on src/classes/ACCT_Accounts_TDTM.cls:
```diff
@@ -39,8 +39,8 @@
 * name updating.
 */
 public class ACCT_Accounts_TDTM extends TDTM_Runnable {
-        
```
No logic change in this class. Adv does not use NPSP's HouseholdNaming logic, so I extract the rest logic into a separate function. So, adv can call it.

weizsforg commented:
Hi @lianeyee I just updated the notes in this PR. Can you please review it? Basically, this PR is about providing support to Adv from NPSP. How detail we need to go about the implementation?

ceiroa commented on src/classes/Advancement_Adapter.cls:
```diff
@@ -0,0 +1,118 @@
+    /**
+     * @description To be called by the NPSP TDTM trigger dispatcher. Advancement will disable some NPSP trigger handlers records (since they conflict with HEDA version)
+     * and return the rest
+     *
```
Missing @param annotation.

ceiroa commented on src/classes/Advancement_Adapter.cls:
```diff
@@ -0,0 +1,118 @@
+     */
+    global virtual List<Trigger_Handler__c> getDefaultTdtmConfigNpspRecords(List<Trigger_Handler__c> listTH) {
+        if(!Advancement_Info.useAdv()) {
+            throw new AdapterException('Invalid call. This function is reserved for Salesforce use.');
```
Same as in the HEDA TDTM PR. We should use a label for the exception messages.

ceiroa commented on src/classes/Advancement_Adapter.cls:
```diff
@@ -0,0 +1,118 @@
+    /**
+     * @description Since Advancement will disable some NPSP trigger handler class, we need to fill in gap in HEDA.
+     * Expose this API for Advancement to inject the moving opportunity functionality into HEDA code base
+     *
```
All these method seem to be missing @param annotations.

weizsforg commented:
Hi @ceiroa , this PR is ready for review. 

I added the @params in all the methods. Switched to use Label for the exception. Added the test class for Advancement_Adapter. 

Thanks

ceiroa commented:
@mpusto This is good to go on our side. I believe you wanted to be @mentioned so you could create a WI to have the item properly reviewed by someone on the NPSP team. For now I have assigned the item to @davidhabib.

Let us know if you have any questions. Thanks!

weizsforg commented on src/classes/ACCT_AccountMerge_TEST.cls:
```diff
@@ -71,6 +71,9 @@ private with sharing class ACCT_AccountMerge_TEST {
     */
     static testMethod void testMergeTwoHHAccounts() {
+        //skip the test if Advancement is installed
+        if(Advancement_Info.useAdv()) return;
```
Basically, the idea is that we don't want to run certain NPSP test cases in Adv since it will fail anyway. 

Even Advancement does pass tdtm configuration back, the test cases still going to fail. The reason is that the NPSP unit test has the unique setup environment and it is not aware about HEDA trigger and HEDA config. And it is tricky (no way?) to setup HEDA trigger and HEDA config for NPSP unit test cases. As a result, some unit test cases will fail due to the conflict with HEDA. 

This is one of the limitation of the current Adv design, we need to do a periodically check when we apply the new version of NPSP and comment out more test case if needed. Any suggestions?

weizsforg commented on src/classes/Advancement_Info.cls:
```diff
@@ -0,0 +1,107 @@
+            Boolean isUserLicensed = UserInfo.isCurrentUserLicensed('adv1');
+            if(isUserLicensed) {
+                List<SObject> advSettings = Database.query('select adv1__Use_Advancement__c, adv1__API_NPSP_Class__c from adv1__Advancement_Setting__mdt');
+                useAdv = (Boolean)advSettings[0].get('adv1__Use_Advancement__c');
```
Yes, I think we need to add the check of the number of records.

weizsforg commented on src/classes/Advancement_Info.cls:
```diff
@@ -0,0 +1,107 @@
+                useAdv = (Boolean)advSettings[0].get('adv1__Use_Advancement__c');
+                String apiNPSPClass = (String)advSettings[0].get('adv1__API_NPSP_Class__c');
+
+                if(String.isNotBlank(apiNPSPClass)) {
```
Right, we need to check if useAdv is false here.

weizsforg commented on src/classes/Advancement_Info.cls:
```diff
@@ -0,0 +1,107 @@
+                    Type classType = Type.forName(apiNPSPClass);
+                    if (classType != null) {
+                        Object classInstance = classType.newInstance();
+                        if (classInstance instanceof Advancement_Adapter) {
```
Thanks! 
We can throw a new exception to the user similar to the BDI_DataImportService.instantiateClassForInterface()

weizsforg commented on src/classes/Advancement_Info.cls:
```diff
@@ -0,0 +1,107 @@
+                }
+            }
+        } catch(Exception e) {
+            // Advancement not installed. Nothing to do
```
Agree! I will update the code to throw exception to the user with proper error message when Adv is installed but have errors in the configuration or licensing.

weizsforg commented on src/classes/Advancement_Info.cls:
```diff
@@ -0,0 +1,107 @@
+    * @return useAdv The boolean value to determine if Adv project is installed.
+    *******************************************************************************************************/
+    public static Boolean useAdv() {
+        if(!setupCalled) {
```
Good point! I can update that.

weizsforg commented on src/classes/Advancement_Info.cls-meta.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexClass xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>41.0</apiVersion>
```
I can update it to API 37

weizsforg commented on src/classes/Advancement_Adapter_TEST.cls-meta.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexClass xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>41.0</apiVersion>
```
Yes, I can update it to API 37.

weizsforg commented on src/classes/Advancement_Adapter.cls-meta.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexClass xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>41.0</apiVersion>
```
Yes, I can update it to API 37

weizsforg commented on src/classes/Advancement_Adapter.cls:
```diff
@@ -0,0 +1,123 @@
+* @description Provides an adapter for Advancement package to be able to inject customizations into NPSP and provide API to inject to other project.
+**/
+
+global virtual with sharing class Advancement_Adapter {
```
Yes, I will update the class name.

weizsforg commented on src/classes/TDTM_ObjectDataGateway.cls:
```diff
@@ -91,49 +91,16 @@ public without sharing class TDTM_ObjectDataGateway implements TDTM_iTableDataGa
-            if (listTH == null) {
-                listTH =  [select Asynchronous__c, Class__c, Load_Order__c, Object__c, Active__c, Usernames_to_Exclude__c, 
-                    Trigger_Action__c from Trigger_Handler__c order by Load_Order__c];
+            if(Advancement_Info.useAdv() && Advancement_Info.getApiNPSP() != null) {
```
I think it is safer to write it in this way.  Basically, Adv need to have the full control about what "listTH" gets return to NPSP if Adv is installed. Even the listTH is not empty, we still pass it into Adv class. And in Adv class will decide whether pass it back or re query it. 

1. Since listTH is a public static variable, it is possible that someone set it directly from the code. If we just listTH if listTH != null, then Adv will lose the ability to filter it. 

2. The code is more clear to other developer. We just say, if Adv is installed then let Adv handle it. Otherwise just the original NPSP code. So, we can be sure that NPSP will not break if Adv is not installed.

davidhabib commented on src/classes/TDTM_ObjectDataGateway.cls:
```diff
@@ -91,49 +91,16 @@ public without sharing class TDTM_ObjectDataGateway implements TDTM_iTableDataGa
-                listTH =  [select Asynchronous__c, Class__c, Load_Order__c, Object__c, Active__c, Usernames_to_Exclude__c, 
-                    Trigger_Action__c from Trigger_Handler__c order by Load_Order__c];
+            if(ADV_Info.useAdv() && ADV_Info.getApiNPSP() != null) {
+                listTH = ADV_Info.getApiNPSP().getDefaultTdtmConfigNpspRecords(listTH);
```
so by having Adv always reset listTH, it is preventing any customer code from modifying listTH in the transaction, which NPSP says it supports in TDTM_Config_API.getCachedRecords().

    * @description Returns the list of current cached trigger handler records
    * @return List<Trigger_Handler__c> a list of cached trigger handler records with current values.  SI's
    * may modify this list in the current execution context to temporarily change the functionality of
    * tdtm triggers - disable triggers, enable triggers, or even add a new row to the collection
    * that is not committed to the database.

is this intentional?  if so, I think we need an explicit comment in the code.

@ceiroa want to make sure you see this discussion.

weizsforg replied:
Actually, no. Adv will not totally reset the listTH. What Adv will do, it to take whatever in the listTH and disable some TriggerHandlers that will conflict with HEDA in order to make sure Adv works correctly. 

1. If customer only have NPSP, then they can do whatever they want since the logic is still the same.

2. If customer have NPSP + Adv, it will not prevent the customer call TDTM_Config_API.getCachedRecords(). Adv will only disable the TriggerHandler that has to be disable and then return the rest back. 

3. If later anything need to be update, we can just update it in Adv package and leave NPSP untouch since the logic is in Adv

I will double check the comments to make sure it makes sense.

ceiroa replied:
Right, at the end the call goes to https://github.com/SalesforceFoundation/adv/pull/2/files#diff-3ea107cf2322681898732395ee4951f2R92, which just deactivates some NPSP TMTD classes. That looks correct to me...

davidhabib commented on src/labels/CustomLabels.labels:
```diff
@@ -1100,6 +1107,13 @@ If you’re not sure how to resolve these errors, post a message in the Nonprofi
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Config Error in Adv</shortDescription>
+        <value>The Gift Processing package is not config correctly. Please contact your Admin for more details.</value>
```
we try to get the correct text from the Doc writer before we merge into master, otherwise it will be missed and forgotten!

weizsforg replied:
Yes, I will check with @lianeyee for this one since @juliesnow is not available today.

lianeyee replied:
@weizsforg Couple of things. Should we avoid referring to "adv" everywhere, not just in the label text itself? So . . .
<fullName>giftProcessingConfigException</fullName>
<shortDescription>Gift Processing configuration error</shortDescription>

I left a comment in the PR conversation looking for more context about when this error appears. Assuming it's more common for an end user to see the error, I suggest:
<value>Your org is missing required configuration for Gift Processing. Contact your admin for help.</value>

weizsforg commented on src/classes/ADV_Info.cls:
```diff
@@ -0,0 +1,127 @@
+        if(isUserLicensed) {
+            List<SObject> advSettings = Database.query('select adv1__Use_Advancement__c, adv1__API_NPSP_Class__c from adv1__Advancement_Setting__mdt');
+            if(advSettings != null && advSettings.size() > 0) {
+                useAdv = (Boolean)advSettings[0].get('adv1__Use_Advancement__c');
```
I think we could clear the hasConfigError here since useAdv can be false if user decided it.

weizsforg commented on src/classes/ADV_Info.cls:
```diff
@@ -0,0 +1,127 @@
+            throw new ADVException(Label.advConfigException);
+        }
+
+        setupCalled = true;
```
Right, I can put it into a different place.

weizsforg commented on src/classes/ADV_Adapter.cls:
```diff
@@ -0,0 +1,122 @@
+    * @return List<Trigger_Handler__c> The list of Trigger Handler record for TDTM to use.
+    ********************************************************************************************************/
+    global virtual List<Trigger_Handler__c> getDefaultTdtmConfigNpspRecords(List<Trigger_Handler__c> listTH) {
+        if(!ADV_Info.useAdv()) {
```
Yes, I will add the space of it.

lianeyee commented:
@weizsforg When or in what context would the incorrect config error be thrown? Should the message be written more for the end user or for the admin? (Picking this up for Julie, who's OOO for a bit today.)

weizsforg commented on src/classes/ADV_Adapter.cls:
```diff
@@ -0,0 +1,122 @@
+        TDTM_Runnable.Action triggerAction = TDTM_TriggerActionHelper.getTriggerActionFromName(action);
+        ACCT_Accounts_TDTM accountsTDTM = new ACCT_Accounts_TDTM();
+        accountsTDTM.configAccountForContact(listNew, listOld, triggerAction, new List<Account>(), new List<Account>());
+    }
```
Yes, it is intentional. In Adv, we have our own HH naming convention and settings that almost the same as NPSP. So, we don't need NPSP to run it again.

weizsforg commented on src/classes/ADV_Adapter.cls:
```diff
@@ -0,0 +1,122 @@
+        }
+
+        a.npe01__SYSTEM_AccountType__c = CAO_Constants.isHHAccountModel() ? CAO_Constants.HH_ACCOUNT_TYPE : CAO_Constants.ONE_TO_ONE_ORGANIZATION_TYPE;
+        a.Type = CAO_Constants.isHHAccountModel() ? CAO_Constants.HH_TYPE : '';
```
hmm. Yes, I can add the check here. Thanks!

weizsforg commented on src/classes/ADV_Adapter_TEST.cls:
```diff
@@ -0,0 +1,128 @@
+        ADV_Adapter theAdapter = new ADV_Adapter();
+        Test.startTest();
+        try {
+            List<Trigger_Handler__c> triggerHandlers = theAdapter.getDefaultTdtmConfigNpspRecords(null);
```
Good point!

weizsforg commented on src/classes/ADV_Adapter_TEST.cls:
```diff
@@ -0,0 +1,128 @@
+        ADV_Adapter theAdapter = new ADV_Adapter();
+        Test.startTest();
+        try {
+            theAdapter.rollupAccounts(null);
```
Good point!

weizsforg commented on src/classes/ADV_Adapter_TEST.cls:
```diff
@@ -0,0 +1,128 @@
+        ADV_Adapter theAdapter = new ADV_Adapter();
+        Test.startTest();
+        try {
+            theAdapter.moveOppsToContactAccount(null, null);
```
Good point!

weizsforg commented on src/classes/ADV_Adapter_TEST.cls:
```diff
@@ -0,0 +1,128 @@
+        ADV_Adapter theAdapter = new ADV_Adapter();
+        Test.startTest();
+        try {
+            theAdapter.configAccountForContact(null);
```
Good point!

weizsforg commented on src/classes/ADV_Adapter_TEST.cls:
```diff
@@ -0,0 +1,128 @@
+        ADV_Adapter theAdapter = new ADV_Adapter();
+        Test.startTest();
+        try {
+            theAdapter.configAccountTDTM(null, null, null);
```
Good point!

weizsforg commented:
@lianeyee There are two use cases for the error message.  1. Gift Processing is installed, but the customer does't have the license to use it. 2 Gift Processing is installed and licensed, but the config in custom metadata in Gift Processing is not set up correctly. For example, the typo in the API class API_HEDA -> API_HEDAA. 

I think the error is more for the first person that use the system. Since if any of the issue happen, system will not work in the first place. But it will also happen when someone accidentally change the Gift Processing settings.

lianeyee commented:
Thanks for the extra context, @weizsforg. Is sounds like the text proposed above will work then. Let us know if you need us to tweak it any further.

weizsforg commented:
Thanks @lianeyee ! I will make the change based your comments.

weizsforg commented:
@lianeyee @juliesnow . We need to add another label for exception.
In NPSP, we created an API for Gift Processing to set up some Account fields based on the Account Model (Household or One-One). If the Account model is Bucket account, we need to throw an exception from the API method. 

Can we have a custom label like:
giftProcessingAccountException
Gift Processing Account Model error
Gift Processing currently does not support Bucket Account Model!

weizsforg commented:
Hi @davidhabib , thanks for the feedback! Can you please review the changes? I will also update the new label after I get feedback from the Doc team.

force2b commented on src/classes/ADV_Adapter.cls:
```diff
@@ -0,0 +1,126 @@
+/*
+    Copyright (c) 2017, Salesforce.org
```
@weizsforg The year in the copyright and apexdoc comment below be 2018 for new classes/triggers.

weizsforg replied:
Oh, Yes. I will update it.

force2b replied:
I see the change for the year in the apex doc comment, but the copyright year is still 2017 on the classes.

force2b commented on src/classes/ADV_Adapter.cls:
```diff
@@ -0,0 +1,126 @@
+    * @param listTH The list of Trigger handler record for further processing.
+    * @return List<Trigger_Handler__c> The list of Trigger Handler record for TDTM to use.
+    ********************************************************************************************************/
+    global virtual List<Trigger_Handler__c> getDefaultTdtmConfigNpspRecords(List<Trigger_Handler__c> listTH) {
```
@weizsforg `TDTM_Config_API.getDefaultRecords()` is already a global exposed api method, so I'm not sure of the value making another global exposed method to call the existing one. Also, I see this method takes a parameter of `List<Trigger_Handler__c>`, but doesn't really do anything with it.

weizsforg replied:
@force2b We made this method as a virtual method. So, in Gift Processing package, we can override it with other logic if needed. By using this wrapper method, Gift Processing can just use one method to have different behavior.  Also, we used the parameter of List<Trigger_Handler__c> in Gift Processing when we override the method.

force2b commented on src/classes/ADV_Adapter.cls:
```diff
@@ -0,0 +1,126 @@
+            throw new AdapterException(Label.adapterException);
+        }
+
+        ACCT_IndividualAccounts_TDTM.rollupAccounts(listAccountId);
```
@weizsforg This won't work with Customizable Rollups. There currently isn't a single method you can call that will execute the appropriate rollup engine for an Id, but after Customizable Rollups is merged into Master we can work with you allow this method to work no matter which engine is active.

weizsforg replied:
@force2b Thanks for the information! Yes, we might need to revisit it. Another thing to mention here is that this method is not a general purpose rollup method that we will use it Gift Processing. We used for a specific purpose.  Basically, we want to achieve the same feature as 
NPSP -> ACCT_IndividualAccounts_TDTM  -> rollupAccounts(list<Id> listAccountId) method. 

So, if the there is no change in NPSP ->  ACCT_IndividualAccounts_TDTM  -> rollupAccounts, then we might not need to change this one.

force2b commented on src/classes/ADV_Info.cls:
```diff
@@ -0,0 +1,131 @@
+* @description Provides the Advancement Information to NPSP class.
+**/
+
+public with sharing class ADV_Info {
```
@weizsforg I know there's a lot of code that references this class name, but it seems a bit short. Something like `ADV_PackageInfo_SVC` (a service class related to the Advancement package) would be a more descriptive class name.

weizsforg replied:
Thanks! I can update the class name.

force2b commented on src/classes/ADV_Info.cls:
```diff
@@ -0,0 +1,131 @@
+    private static void setup() {
+
+        //Check if setup() is being called. If yes, just return
+        if (setupCalled)
```
@weizsforg  The general coding style in use throughout NPSP (at least anything built in recent years) uses opening and closing curly brackets on all code blocks even if they're just a single line:
```
if (setupCalled) {
   return;
}
```

weizsforg replied:
Thanks. I can add the curly brackets.

force2b commented on src/classes/ACCT_AccountMerge_TEST.cls:
```diff
@@ -71,6 +71,9 @@ private with sharing class ACCT_AccountMerge_TEST {
     */
     static testMethod void testMergeTwoHHAccounts() {
+        //skip the test if Advancement is installed
+        if(ADV_PackageInfo_SVC.useAdv()) return;
```
@weizsforg Thanks for making this change. Been thinking about this code pattern all day. Seems to me that we're basically hardcoding all of our unit tests to very specifically check for the Advancement package. Though that meets the current need it doesn't extend beyond Advancement. So, rather than check for `if(ADV_PackageInfo_SVC.useAdv())` I think it might be better to create a new more generic class in NPSP for `UTIL_PackageExtension_SVC` that has a method for `skipUnitTests()`. For now that method can return the value of `ADV_PackageInfo_SVC.useAdv()`, but it would allow us to modify a single bit of code to expand this usage beyond the Advancement package.

In addition, it might be helpful if `skipUnitTests()` could accept an Enum parameter to know which types of unit tests to skip. Some possible examples are:
```
UTIL_PackageExtension_SVC.skipUnitTests(UTIL_PackageExtension_SVC.TestType.Household);
UTIL_PackageExtension_SVC.skipUnitTests(UTIL_PackageExtension_SVC.TestType.Rollups);
UTIL_PackageExtension_SVC.skipUnitTests(UTIL_PackageExtension_SVC.TestType.Allocations);
UTIL_PackageExtension_SVC.skipUnitTests(UTIL_PackageExtension_SVC.TestType.Addresses);
```

I'm happy to talk through this more.

ceiroa replied:
Hey @force2b, any chance we could continue this way and open a technical debt item in our backlog to address this at a later date? We have been working on this literally for moths (the PR with the other side of this was opened in December), and would like to get it in as soon as we can, provided it meets all technical requirements. :-)

davidhabib replied:
I agree @ceiroa.  for now let's leave it as is.  @force2b I think this is something the NPSP team could consider doing in the future once they better define it.  But I don't think we should be asking Adv to do this at this time.

force2b replied:
Completely understand @ceiroa. I'm coming in a bit late on this. My primary concern is extensibility as we add support for additional external packages (thinking of the Archival Bridge package). Still, I think we're fine with this architecture in the short term. I don't see it creating any packaging issues if refactoring is required at a later point.

weizsforg commented:
Hi @lianeyee , any idea about the new label that I mentioned before?

lianeyee commented:
Sorry @weizsforg! Thanks for poking me. I forgot about this one.

giftProcessingAccountException
Gift Processing account model error
Gift Processing doesn't support the Individual ("Bucket") account model.

force2b commented:
@ceiroa and @weizsforg Now that this is merged into Master and will be sync'd up to all of our feature branches, the NGO team will need guidance on when to implement the `if(ADV_PackageInfo_SVC.useAdv()) return;` code in the new unit test classes that we're building. For example, there are 20 new Unit Test Classes in Customizable Rollups. I'm sure the Soft Credit Automation adds a handful as well.

weizsforg commented:
Thanks @davidhabib @force2b @ceiroa for the help! 
We had a discussion about how to handle the new test cases.  We are planning to add a "Feature Test - Gift Processing " as a new check in the PR. So, it will tell NPSP team know which test class failed under gift processing environment. And maybe NPSP team can @ us to have a look and figure out which ones are need to be disabled. For now, we can check in Gift Processing environment and create a new PR in NPSP to disable additional test cases.
